### PR TITLE
Tesla

### PR DESCRIPTION
**Pull request by jgittler**
Move underlying http package to Tesla from HTTPoison.
Move handle response, oauth, and retrying to middleware.
Clean up test placeholders and mock client.

### CODE REVIEW

File: lib/procore/http_client.ex [hunk 1]
```diff
@@ -1,76 +1,10 @@
...
-    if List.keymember?(headers, "application/json", 1),
-      do: ResponseResult.parse_json_body(body),
-      else: body
-  end
+  use Tesla, only: [:post, :get, :patch]
+  plug(Tesla.Middleware.EncodeJson, engine: Poison)
+  plug(Tesla.Middleware.HandleResponse)
+  plug(Tesla.Middleware.DecodeJson, engine: Poison)
+  plug(Tesla.Middleware.RetryWithBackoff)
+  plug(Tesla.Middleware.Logger)
```
Comment (terrence2) [inline lib/procore/http_client.ex]: :tada:

File: lib/tesla/middleware/retry_with_backoff.ex [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  end
+
+  defp retry(env, next, _delay, retries, _max_retries) when retries <= 1 do
+    Tesla.run(env, next)
+  end
+
+  defp retry(env, next, delay, retries, max_retries) do
+    case Tesla.run(env, next) do
+      {:error, reason} ->
+        IO.inspect("RETRYING DUE TO: #{reason}")
```
Comment (terrence2) [inline lib/tesla/middleware/retry_with_backoff.ex]: We should probably use logger for these prints.

File: lib/tesla/middleware/retry_with_backoff.ex [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+      {:ok, env} ->
+        {:ok, env}
+    end
+  end
+
+  defp backoff(delay, retries, max_retries) do
+    :math.pow(delay, max_retries - retries)
+    |> trunc()
+    |> :timer.sleep()
```
Comment (terrence2) [inline lib/tesla/middleware/retry_with_backoff.ex]: There is a Process.sleep now, which I think is the new canonical way to sleep.

File: lib/procore/http_client/oauth.ex [hunk 4]
```diff
@@ -1,62 +1,59 @@
...
   end
+
+  defp tesla_client do
+    Tesla.build_client([{Tesla.Middleware.Headers, headers()}])
+  end
+
+  defp headers do
+    [{"Content-Type", "application/json"}]
+  end
 end
```
Comment (terrence2) [inline lib/procore/http_client/oauth.ex]: I wrote a plug to do oauth that might be more elegant than this, assuming it actually fits into the existing architecture. If you are interested, you can find it here: https://github.com/procore/project-mirror/blob/master/lib/project_mirror/procore/oauth_middleware.ex

Comment (jgittler) [reply lib/procore/http_client/oauth.ex]: Will definitely take a look at this and fit it into the sdk. We haven't focused much on auth, so I'm sure the middleware will be an improvement.

**Pull request merged**